File: typescript/src/astro/nebulae.ts
/**
 * Nebula catalogue types and proximity queries — the **data-free** core of the
 * nebulae feature.
 *
 * Elite Dangerous models thousands of nebulae. This module holds the
 * {@link Nebula} record shape and the pure functions that search a catalogue
 * ({@link nearestNebulae}, {@link nebulaeWithin}, {@link getNebulaByName}); the
 * catalogues themselves live in sibling modules, one per class, so you only bundle
 * the ones you ask for:
 *
 * Sizes are the published minified ESM, before any transport compression; the gzipped
 * figure is roughly a fifth of it.
 *
 * | Module | Export | Entries | Minified | Gzipped |
 * | --- | --- | --- | --- | --- |
 * | `./nebulae-real` | `REAL_NEBULAE` | 180 | 19 KB | 5 KB |
 * | `./nebulae-procgen` | `PROCGEN_NEBULAE` | 166 | 19 KB | 6 KB |
 * | `./nebulae-planetary` | `PLANETARY_NEBULAE` | 5489 | 645 KB | 140 KB |
 * | `./nebulae-all` | `ALL_NEBULAE` | 5835 | 682 KB | 151 KB |
 *
 * Importing a query function from here costs nothing but the function: pass in
 * whichever catalogue you imported. The catalogue argument is **required** — 94% of
 * `ALL_NEBULAE` is planetary nebulae most apps never touch, so there is no default
 * worth falling back to silently.
 *
 * Positions are galactic light-years with Sol at the origin — the same frame the
 * journal, EDSM and Spansh use. Catalogue data originates from EDAstro
 * (CMDR Orvidius); see [`data/astro/SOURCES.md`](https://github.com/DarkSession/Elite-Dangerous-Almanac/blob/main/data/astro/SOURCES.md).
 *
 * @example
 * ```ts
 * import { nearestNebulae } from '@elite-dangerous-almanac/core/astro/nebulae';
 * import { REAL_NEBULAE } from '@elite-dangerous-almanac/core/astro/nebulae-real';
 *
 * nearestNebulae({ x: 0, y: 0, z: 0 }, REAL_NEBULAE, 1)[0].name; // -> 'Pleiades'
 * ```
 *
 * @packageDocumentation
 */

import type { GalacticPosition } from './galactic-position.js';

/**
 * How a nebula is classified by the source catalogue.
 *
 * - `real` — a catalogued real-world nebula or dark region the game models by name
 *   (Horsehead Nebula, Barnard's Loop, Coalsack Dark Region, …).
 * - `planetary` — a planetary nebula, catalogued at the procedurally-named system
 *   it surrounds. By far the largest class.
 * - `procgen` — a procedurally generated nebula, catalogued at a nearby system.
 */
export type NebulaType = 'real' | 'planetary' | 'procgen';

/**
 * One catalogued nebula and where it sits in the galaxy.
 *
 * @remarks
 * A nebula is a volume, but the catalogue records a single reference point — the
 * position of {@link Nebula.system}, the system it is catalogued at. Distances
 * computed from it are therefore "distance to the catalogued system", not distance
 * to the nebula's edge.
 */
export interface Nebula {
    /**
     * Display name, e.g. `"Witch Head Nebula"`. For procedurally-named nebulae this
     * is usually the system name, sometimes with a community name appended in
     * parentheses (`"Aemonz WZ-Y e6771 (Lazurite Nebula)"`).
     */
    readonly name: string;
    /** The in-game system the nebula is catalogued at, e.g. `"Witch Head Sector IR-W c1-8"`. */
    readonly system: string;
    /** Galactic X of the catalogued system, in light-years (Sol at origin). */
    readonly x: number;
    /** Galactic Y of the catalogued system, in light-years (Sol at origin). */
    readonly y: number;
    /** Galactic Z of the catalogued system, in light-years (Sol at origin). */
    readonly z: number;
    /** Which class of nebula this is. */
    readonly type: NebulaType;
    /**
     * Id of the galactic codex region the nebula sits in, 1–42.
     *
     * @remarks
     * Resolve it to a name with `getCodexRegion` from `./codex-region` — that
     * costs ~9 KB of region metadata rather than the ~267 KB lookup grid
     * `findCodexRegionAt` needs.
     */
    readonly regionId: number;
}

/** A {@link Nebula} annotated with its distance from a queried point. */
export interface NebulaWithDistance extends Nebula {
    /** Straight-line distance from the queried coordinates, in light-years. */
    readonly distanceLy: number;
}

/** Squared distance between a point and a nebula's catalogued system, in ly². */
function distanceSquared(coords: GalacticPosition, nebula: Nebula): number {
    const dx = coords.x - nebula.x;
    const dy = coords.y - nebula.y;
    const dz = coords.z - nebula.z;
    return dx * dx + dy * dy + dz * dz;
}

/**
 * The nebulae closest to a point, nearest first.
 *
 * @param coords - The point to measure from, in light-years (Sol at origin). A
 * `ProceduralSystem.position` value fits once you have narrowed it — it is
 * `GalacticPosition | null`, and is `null` unless that system was built from an address
 * *and* you supplied its coordinates, so null-check it first.
 * @param nebulae - The catalogue to search — `REAL_NEBULAE`, `PLANETARY_NEBULAE`,
 * `PROCGEN_NEBULAE`, `ALL_NEBULAE`, or any subset you have filtered yourself.
 * @param count - How many to return. Defaults to `3`. Values `<= 0` yield an empty
 * array; a `count` larger than the catalogue returns the whole catalogue.
 * @returns Up to `count` nebulae sorted by ascending {@link NebulaWithDistance.distanceLy}.
 * Ties keep catalogue order. The input array is not modified.
 * @example
 * ```ts
 * // The three real nebulae nearest Sol
 * nearestNebulae({ x: 0, y: 0, z: 0 }, REAL_NEBULAE).map((n) => n.name);
 * // -> [ 'Pleiades', 'R Cra', 'Lupus Dark Region B' ]  (nearest first)
 *
 * nearestNebulae({ x: 0, y: 0, z: 0 }, REAL_NEBULAE, 1)[0].distanceLy; // -> ≈383.31
 * ```
 */
export function nearestNebulae(
    coords: GalacticPosition,
    nebulae: readonly Nebula[],
    count = 3,
): NebulaWithDistance[] {
    if (count <= 0) return [];
    // Rank on squared distances (no sqrt per comparison), then take the square root
    // only for the handful of records actually returned.
    const ranked = nebulae.map((nebula) => ({ nebula, d2: distanceSquared(coords, nebula) }));
    ranked.sort((a, b) => a.d2 - b.d2);
    return ranked
        .slice(0, count)
        .map(({ nebula, d2 }) => ({ ...nebula, distanceLy: Math.sqrt(d2) }));
}

/**
 * Every nebula within a radius of a point, nearest first.
 *
 * @param coords - The point to measure from, in light-years (Sol at origin).
 * @param nebulae - The catalogue to search (see {@link nearestNebulae}).
 * @param radiusLy - Search radius in light-years, inclusive of the boundary.
 * Negative radii yield an empty array.
 * @returns The matching nebulae sorted by ascending
 * {@link NebulaWithDistance.distanceLy}. Ties keep catalogue order. The input array
 * is not modified.
 * @example
 * ```ts
 * // Real nebulae within 400 ly of Sol
 * nebulaeWithin({ x: 0, y: 0, z: 0 }, REAL_NEBULAE, 400).map((n) => n.name);
 * // -> [ 'Pleiades' ]
 * ```
 */
export function nebulaeWithin(
    coords: GalacticPosition,
    nebulae: readonly Nebula[],
    radiusLy: number,
): NebulaWithDistance[] {
    if (radiusLy < 0) return [];
    const limit = radiusLy * radiusLy;
    const hits: NebulaWithDistance[] = [];
    for (const nebula of nebulae) {
        const d2 = distanceSquared(coords, nebula);
        if (d2 <= limit) hits.push({ ...nebula, distanceLy: Math.sqrt(d2) });
    }
    hits.sort((a, b) => a.distanceLy - b.distanceLy);
    return hits;
}

/**
 * Look up a nebula by name (case-insensitive).
 *
 * @param name - The nebula name as the catalogue spells it, e.g.
 * `"Witch Head Nebula"`. Matching ignores case and surrounding whitespace but is
 * otherwise exact — including any parenthetical community name
 * (`"Aemonz WZ-Y e6771 (Lazurite Nebula)"`), so pass the whole string.
 * @param nebulae - The catalogue to search (see {@link nearestNebulae}).
 * @returns The matching {@link Nebula}, or `null` if the catalogue holds no nebula
 * of that name. Only the catalogue you pass is searched — a real nebula will not
 * be found in `PLANETARY_NEBULAE`.
 * @example
 * ```ts
 * getNebulaByName('witch head nebula', REAL_NEBULAE)?.system;
 * // -> 'Witch Head Sector RY-R b4-0'
 * ```
 */
export function getNebulaByName(name: string, nebulae: readonly Nebula[]): Nebula | null {
    const wanted = name.trim().toLowerCase();
    return nebulae.find((nebula) => nebula.name.toLowerCase() === wanted) ?? null;
}

File: dotnet/src/EliteDangerousAlmanac/Astronomy/Nebulae.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using EliteDangerousAlmanac.Internal;

namespace EliteDangerousAlmanac.Astronomy;

/// <summary>How the catalogue classes one nebula.</summary>
public enum NebulaType
{
    /// <summary>A real nebula or dark region the game models under its own name.</summary>
    Real,

    /// <summary>A planetary nebula, catalogued at the system it surrounds.</summary>
    Planetary,

    /// <summary>A nebula the generator made, catalogued at a system beside it.</summary>
    Procgen,
}

/// <summary>One catalogued nebula, and where in the galaxy it sits.</summary>
/// <param name="Name">
/// The display name. A nebula the generator made usually carries its system's name, and
/// sometimes a community name after it in brackets.
/// </param>
/// <param name="System">The system the nebula is catalogued at.</param>
/// <param name="X">The system's galactic X in light years, with Sol at the origin.</param>
/// <param name="Y">The system's galactic Y in light years.</param>
/// <param name="Z">The system's galactic Z in light years.</param>
/// <param name="Type">Which class of nebula this is.</param>
/// <param name="RegionId">
/// The number of the codex region the nebula sits in, one through 42, which
/// <see cref="CodexRegions.Find"/> reads.
/// </param>
/// <remarks>
/// A nebula fills space, and the catalogue records one point of it: where its system is.
/// A distance from that point is therefore a distance to the system, and not to the edge
/// of the cloud.
/// </remarks>
public record Nebula(
    string Name,
    string System,
    double X,
    double Y,
    double Z,
    NebulaType Type,
    int RegionId);

/// <summary>One catalogued nebula, and how far it is from a place that was asked about.</summary>
/// <param name="Name">The display name.</param>
/// <param name="System">The system the nebula is catalogued at.</param>
/// <param name="X">The system's galactic X in light years, with Sol at the origin.</param>
/// <param name="Y">The system's galactic Y in light years.</param>
/// <param name="Z">The system's galactic Z in light years.</param>
/// <param name="Type">Which class of nebula this is.</param>
/// <param name="RegionId">The number of the codex region the nebula sits in.</param>
/// <param name="DistanceLy">The straight-line distance from the place asked about, in light years.</param>
public sealed record NebulaNearby(
    string Name,
    string System,
    double X,
    double Y,
    double Z,
    NebulaType Type,
    int RegionId,
    double DistanceLy)
    : Nebula(Name, System, X, Y, Z, Type, RegionId);

/// <summary>The nebula catalogues, and the searches that read them.</summary>
/// <remarks>
/// <para>
/// The game models thousands of nebulae, in three classes. <see cref="Real"/> holds the
/// real nebulae and dark regions a commander recognises, and it is the one most callers
/// want. <see cref="Planetary"/> is by far the largest, and <see cref="Procgen"/> holds
/// the ones the generator made. <see cref="All"/> is the three of them together.
/// </para>
/// <para>
/// Every search takes the catalogue to read. Nearly all of <see cref="All"/> is planetary
/// nebulae that most callers never touch, so there is no one catalogue worth reading
/// without being asked to.
/// </para>
/// <para>
/// Positions are light years with Sol at the origin, which is the frame a journal, EDSM
/// and Spansh report. The catalogue comes from EDAstro. See <c>ATTRIBUTIONS.md</c> for
/// credit and licence terms.
/// </para>
/// </remarks>
public static class NebulaCatalogue
{
    private static readonly Lazy<IReadOnlyList<Nebula>> RealNebulae = new(
        () => Build("data/astro/nebulae-real.jsonc", NebulaType.Real));

    private static readonly Lazy<IReadOnlyList<Nebula>> PlanetaryNebulae = new(
        () => Build("data/astro/nebulae-planetary.jsonc", NebulaType.Planetary));

    private static readonly Lazy<IReadOnlyList<Nebula>> ProcgenNebulae = new(
        () => Build("data/astro/nebulae-procgen.jsonc", NebulaType.Procgen));

    private static readonly Lazy<IReadOnlyList<Nebula>> AllNebulae = new(BuildAll);

    /// <summary>Every real nebula and dark region the game names, in order of name.</summary>
    public static IReadOnlyList<Nebula> Real => RealNebulae.Value;

    /// <summary>Every planetary nebula, in order of name.</summary>
    public static IReadOnlyList<Nebula> Planetary => PlanetaryNebulae.Value;

    /// <summary>Every nebula the generator made, in order of name.</summary>
    public static IReadOnlyList<Nebula> Procgen => ProcgenNebulae.Value;

    /// <summary>Every catalogued nebula: the real, the planetary and the generated ones.</summary>
    /// <remarks>Read <see cref="Nebula.Type"/> to narrow it down afterwards.</remarks>
    public static IReadOnlyList<Nebula> All => AllNebulae.Value;

    /// <summary>Reads the nebulae nearest one place, nearest first.</summary>
    /// <param name="position">The place to measure from, in light years, with Sol at the origin.</param>
    /// <param name="nebulae">The catalogue to read.</param>
    /// <param name="count">
    /// How many to answer with. A count of zero or below answers with none, and a count
    /// past the catalogue answers with the whole of it.
    /// </param>
    /// <returns>
    /// Up to <paramref name="count"/> nebulae, in order of distance. Two at the same
    /// distance keep the order the catalogue holds them in. The catalogue is left alone.
    /// </returns>
    /// <exception cref="ArgumentNullException">The place or the catalogue are absent.</exception>
    public static IReadOnlyList<NebulaNearby> Nearest(
        GalacticPosition position, IReadOnlyList<Nebula> nebulae, int count = 3)
    {
        if (position is null) throw new ArgumentNullException(nameof(position));
        if (nebulae is null) throw new ArgumentNullException(nameof(nebulae));

        int limit = Math.Min(Math.Max(count, 0), nebulae.Count);
        if (limit == 0) return [];

        // Only the nearest few are kept, on a heap whose root is the furthest of them.
        // The whole catalogue is therefore never ranked or ordered for the usual small
        // question, and two at the same distance still keep catalogue order.
        List<Ranked> nearest = new(limit);
        for (int index = 0; index < nebulae.Count; index++)
        {
            Ranked ranked = new(nebulae[index], DistanceSquared(position, nebulae[index]), index);
            if (nearest.Count < limit)
            {
                PushFurthest(nearest, ranked);
            }
            else if (Compare(ranked, nearest[0]) < 0)
            {
                ReplaceFurthest(nearest, ranked);
            }
        }

        nearest.Sort(Compare);
        NebulaNearby[] answer = new NebulaNearby[nearest.Count];
        for (int index = 0; index < nearest.Count; index++) answer[index] = WithDistance(nearest[index]);
        return new ReadOnlyCollection<NebulaNearby>(answer);
    }

    /// <summary>Reads every nebula within a distance of one place, nearest first.</summary>
    /// <param name="position">The place to measure from, in light years, with Sol at the origin.</param>
    /// <param name="nebulae">The catalogue to read.</param>
    /// <param name="radiusLy">
    /// How far to reach, in light years, counting one exactly that far away. A distance
    /// below zero answers with none.
    /// </param>
    /// <returns>
    /// The nebulae that near, in order of distance. Two at the same distance keep the
    /// order the catalogue holds them in. The catalogue is left alone.
    /// </returns>
    /// <exception cref="ArgumentNullException">The place or the catalogue are absent.</exception>
    public static IReadOnlyList<NebulaNearby> Within(
        GalacticPosition position, IReadOnlyList<Nebula> nebulae, double radiusLy)
    {
        if (position is null) throw new ArgumentNullException(nameof(position));
        if (nebulae is null) throw new ArgumentNullException(nameof(nebulae));
        if (!(radiusLy >= 0)) return [];

        double limit = radiusLy * radiusLy;
        List<Ranked> hits = [];
        for (int index = 0; index < nebulae.Count; index++)
        {
            double distance = DistanceSquared(position, nebulae[index]);
            if (distance <= limit) hits.Add(new Ranked(nebulae[index], distance, index));
        }

        hits.Sort(Compare);
        NebulaNearby[] answer = new NebulaNearby[hits.Count];
        for (int index = 0; index < hits.Count; index++) answer[index] = WithDistance(hits[index]);
        return new ReadOnlyCollection<NebulaNearby>(answer);
    }

    /// <summary>Finds a nebula by name in one catalogue.</summary>
    /// <param name="name">
    /// The name as the catalogue writes it. Any community name in brackets is part of it,
    /// so hand over the whole string.
    /// </param>
    /// <param name="nebulae">The catalogue to read.</param>
    /// <returns>
    /// The nebula, or <see langword="null"/> where that catalogue holds no nebula of the
    /// name. Only the catalogue handed over is read, so a real nebula is not found among
    /// the planetary ones.
    /// </returns>
    /// <exception cref="ArgumentNullException">The catalogue is absent.</exception>
    /// <remarks>Case and surrounding whitespace are ignored, and nothing else is.</remarks>
    public static Nebula? FindByName(string? name, IReadOnlyList<Nebula> nebulae)
    {
        if (nebulae is null) throw new ArgumentNullException(nameof(nebulae));
        return RegistryIndex.FindByKey(nebulae, nebula => nebula.Name, name);
    }

    /// <summary>Reads how far one nebula is from a place, squared.</summary>
    private static double DistanceSquared(GalacticPosition position, Nebula nebula)
    {
        double dx = position.X - nebula.X;
        double dy = position.Y - nebula.Y;
        double dz = position.Z - nebula.Z;
        return (dx * dx) + (dy * dy) + (dz * dz);
    }

    /// <summary>Orders two ranked nebulae by distance, then by catalogue order.</summary>
    private static int Compare(Ranked left, Ranked right)
    {
        int byDistance = left.DistanceSquared.CompareTo(right.DistanceSquared);
        return byDistance != 0 ? byDistance : left.CatalogueIndex.CompareTo(right.CatalogueIndex);
    }

    /// <summary>Writes one ranked nebula with the distance it was ranked at.</summary>
    private static NebulaNearby WithDistance(Ranked ranked) => new(
        ranked.Nebula.Name,
        ranked.Nebula.System,
        ranked.Nebula.X,
        ranked.Nebula.Y,
        ranked.Nebula.Z,
        ranked.Nebula.Type,
        ranked.Nebula.RegionId,
        Math.Sqrt(ranked.DistanceSquared));

    /// <summary>Adds one nebula to the heap whose root is the furthest kept.</summary>
    private static void PushFurthest(List<Ranked> heap, Ranked ranked)
    {
        heap.Add(ranked);
        int index = heap.Count - 1;
        while (index > 0)
        {
            int parent = (index - 1) / 2;
            if (Compare(heap[parent], heap[index]) >= 0) return;
            (heap[parent], heap[index]) = (heap[index], heap[parent]);
            index = parent;
        }
    }

    /// <summary>Puts one nebula in place of the furthest kept, and settles the heap.</summary>
    private static void ReplaceFurthest(List<Ranked> heap, Ranked ranked)
    {
        heap[0] = ranked;
        int index = 0;
        while (true)
        {
            int left = (index * 2) + 1;
            if (left >= heap.Count) return;
            int right = left + 1;
            int further = right < heap.Count && Compare(heap[right], heap[left]) > 0 ? right : left;
            if (Compare(heap[index], heap[further]) >= 0) return;
            (heap[index], heap[further]) = (heap[further], heap[index]);
            index = further;
        }
    }

    /// <summary>Reads one catalogue, and gives every record the class its file states.</summary>
    private static ReadOnlyCollection<Nebula> Build(string path, NebulaType type)
    {
        IReadOnlyList<NebulaRecord> records = SharedData.LoadList<NebulaRecord>(path);
        List<Nebula> nebulae = new(records.Count);
        foreach (NebulaRecord record in records)
        {
            nebulae.Add(new Nebula(
                record.Name,
                record.System ?? record.Name,
                record.X,
                record.Y,
                record.Z,
                type,
                record.RegionId));
        }

        return new ReadOnlyCollection<Nebula>(nebulae);
    }

    /// <summary>Puts the three catalogues together, in the order they are named in.</summary>
    private static ReadOnlyCollection<Nebula> BuildAll()
    {
        List<Nebula> nebulae = new(Real.Count + Planetary.Count + Procgen.Count);
        nebulae.AddRange(Real);
        nebulae.AddRange(Planetary);
        nebulae.AddRange(Procgen);
        return new ReadOnlyCollection<Nebula>(nebulae);
    }

    /// <summary>One nebula as its shared file states it.</summary>
    /// <remarks>
    /// The file settles the class, so no record repeats it. A planetary nebula states its
    /// system only where that differs from its name.
    /// </remarks>
    private sealed record NebulaRecord(
        string Name,
        double X,
        double Y,
        double Z,
        int RegionId,
        string? System = null);

    /// <summary>One nebula, how far away it was found to be, and where the catalogue holds it.</summary>
    private readonly struct Ranked(Nebula nebula, double distanceSquared, int catalogueIndex)
    {
        /// <summary>The nebula itself.</summary>
        public Nebula Nebula { get; } = nebula;

        /// <summary>How far it is from the place asked about, squared.</summary>
        public double DistanceSquared { get; } = distanceSquared;

        /// <summary>Where the catalogue holds it, which settles a tie.</summary>
        public int CatalogueIndex { get; } = catalogueIndex;
    }
}
